api/src/main/java/org/datacleaner/descriptors/ComponentDescriptor.java

```diff
@@ -44,6 +45,13 @@
 public interface ComponentDescriptor<B> extends Comparable<ComponentDescriptor<?>>, Serializable {
 
     /**
+     * It updates component's metadata.
+     * 
+     * @param metadata
+     */
+    public void updateMetadata(Map<String, String> metadata);
```

Comment:
I think the location of this method is a bit difficult... First of all it leaves me wondering "when does this happen". But after answering that (it happens when the component is created) then I think that the order of control is a bit reversed here. A descriptor is usually unaware of the component instances, and I think it's best to keep it that way so that we don't introduce coupling everywhere. The other problem I see is that we don't want the descriptor to just update _anything_ in the map. We want to just provide one specific property, so passing the map around might be dangerous.

My suggestion would be to write and read this metadata as part of the job building API. I would suggest to simply do a check when the component is created: Find the appropriate descriptor provider, and if it's remote then set the metadata property which deals with the origin of the descriptor. That way we also don't pass around a map that could contain _anything_  - we only set one specific metadata property which could/should also be defined as a constant somewhere so that we can do an easy and safe lookup.